File: 95.不同的二叉搜索树-ii.py
#
# @lc app=leetcode.cn id=95 lang=python3
#
# [95] 不同的二叉搜索树 II
#
# https://leetcode-cn.com/problems/unique-binary-search-trees-ii/description/
#
# algorithms
# Medium (62.42%)
# Likes:    390
# Dislikes: 0
# Total Accepted:    30.4K
# Total Submissions: 48.6K
# Testcase Example:  '3'
#
# 给定一个整数 n，生成所有由 1 ... n 为节点所组成的二叉搜索树。
#
# 示例:
#
# 输入: 3
# 输出:
# [
# [1,null,3,2],
# [3,2,null,1],
# [3,1,null,null,2],
# [2,1,3],
# [1,null,2,null,3]
# ]
# 解释:
# 以上的输出对应以下 5 种不同结构的二叉搜索树：
#
# ⁠  1         3     3      2      1
# ⁠   \       /     /      / \      \
# ⁠    3     2     1      1   3      2
# ⁠   /     /       \                 \
# ⁠  2     1         2                 3
#
#
#
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# @lc code=start
# Definition for a binary tree node.


class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def build(start, end):
            ans = []
            if start > end:
                return [None]
            if start == end:
                return [TreeNode(start)]
            for i in range(start, end+1):
                for j in build(start, i-1):
                    for k in build(i+1, end):
                        n = TreeNode(i)
                        n.left = j
                        n.right = k
                        ans.append(n)
            return ans
        return build(1, n)


# @lc code=end

File: 95.不同的二叉搜索树-ii.go
/*
 * @lc app=leetcode.cn id=95 lang=golang
 *
 * [95] 不同的二叉搜索树 II
 *
 * https://leetcode.cn/problems/unique-binary-search-trees-ii/description/
 *
 * algorithms
 * Medium (73.09%)
 * Likes:    1371
 * Dislikes: 0
 * Total Accepted:    157.1K
 * Total Submissions: 215K
 * Testcase Example:  '3'
 *
 * 给你一个整数 n ，请你生成并返回所有由 n 个节点组成且节点值从 1 到 n 互不相同的不同 二叉搜索树 。可以按 任意顺序 返回答案。
 *
 *
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：n = 3
 * 输出：[[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]
 *
 *
 * 示例 2：
 *
 *
 * 输入：n = 1
 * 输出：[[1]]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 *
 *
 *
 *
 */
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func generateTrees(n int) []*TreeNode {
	var dfs func(start, end int) []*TreeNode
	dfs = func(start, end int) []*TreeNode {
		if start > end {
			return []*TreeNode{nil}
		}
		if start == end {
			return []*TreeNode{&TreeNode{Val: start}}
		}
		ans := []*TreeNode{}
		for i := start; i <= end; i++ {
			for _, j := range dfs(start, i-1) {
				for _, k := range dfs(i+1, end) {
					ans = append(ans, &TreeNode{Val: i, Left: j, Right: k})
				}
			}
		}
		return ans
	}
	return dfs(1, n)
}

// @lc code=end
